exchange/src/exchange/tickers/manager/ticker_manager.hpp

```diff
@@ -13,9 +14,28 @@ using engine_ref_t = std::reference_wrapper<nutc::matching::Engine>;
 namespace nutc {
 namespace engine_manager {
 
+enum class EngineState { BOT, RMQ };
+
 class EngineManager : public nutc::ticks::TickObserver {
 public:
-    std::optional<engine_ref_t> get_engine(const std::string& ticker);
+    std::optional<engine_ref_t>
+    get_engine(const std::string& ticker, EngineState require_state)
+    {
+        while (engine_state.load(std::memory_order_acquire) != require_state) {
+            std::this_thread::yield();
+        }
```

Comment:
No. You either use a condition variable (mutex), or a spinlock. Yielding while you spin leads to (significantly) lower perf since the OS swaps your thread out and then you have to wait milliseconds for it to put it back.

Reply:
Disagree. We only have two threads, and one can access the engine at a time. The rationale is that they operate in intervals, so for example:

Time 0-10: RMQ mode, RMQ can make requests without waiting
Time 11-14: Bot mode/on_tick, RMQ has to wait, so instead we pass control to the other thread (tick manager)
Time 15: Bot mode releases/switches mode to RMQ, cycle repeats

However, this implementation is incorrect. This function returns a reference, which we can use past the lifetime of this function. That means we could have two threads using it at once. Example:

Time 0: RMQ requests and gets engine
Time 1: Tick manager(TM) switches mode, requests and gets engine
Time 2: RMQ thread runs operations on the reference
Time 3: TM thread runs operations on the reference

This shouldn't be possible as both are performing operations during bot mode

Reply:
I agree, also this implementation is incorrect. This function returns a reference, which we can use past the lifetime of this function. That means we could have two threads using it at once. Example:

Time 0: RMQ requests and gets engine
Time 1: Tick manager(TM) switches mode, requests and gets engine
Time 2: RMQ thread runs operations on the reference
Time 3: TM thread runs operations on the reference

This shouldn't be possible as both are performing operations during bot mode

You should busy wait and pin each thread to a different core